### ISSUE TITLE
`Hyperlink` datatype

### ISSUE THREAD

**Issue by JosephBond** [explorable-viz/fluid#1206]
We need to include a new datatype, for now we will call it `Hyperlink`, which packages up a value with a string that references or “explains” it.  We need to be able to splice these hyperlink values into a `LinkedText`, so that rather than a list of strings, a  `LinkedText` will instead be a list of `String + Hyperlink` (perhaps wrapped in a newtype). Since we need a way of distinguishing an inert string from a `Hyperlink`, we will wrap the inert strings in their own constructor. Since this will likely get clunky, we may soon want to push on #1187 to streamline things for document authors.

 Since these strings can themselves be “computed”  (the use-case we have investigated so far), interacting with the text once it’s rendered should modify selection information on the (potentially computed, potentially inert) string, and the value to which it makes reference

### Related Issues
- #1187

**Comment by rolyp** [comment]
Suggestions for now:
- [x] Paragraph
  - [x] Rename `LinkedText` to `Paragraph`
  - [x] which takes a list of `TextFragment`
- [x] New `TextFragment` datatype with two constructors:
  - [x] `Link` (with two arguments)
  - [x] `Text` (with one)
- [x] Remove `Explanation` experiment

Other partially-formed thoughts:
- Is the “very likely” pattern better thought of as an example of `Link` where there is both a computed value (e.g. proportion of distribution that is above 2°C) and some computed text (mapping that value to the string “very likely”)? It might be preferable to think of this as a `Link` rather than `Text` because one could easily imagine wanting to be able to reveal the underlying value in the UI. (Although that behaviour would in some sense be redundant if we could visualise the computational explanation behind a value, e.g. a `Text` element.)
- Should a `Link` be a value paired with a function which will interpret that value as a string? Then `Text (numToStr n)` could be expressed as `Link n numToStr` and in fact `Text e` would be equivalent to `Link e id`.

**Comment by JosephBond** [comment]
Are we intending for the `Text` constructor to appear directly in the Fluid syntax?
My next refactoring was planning on changing things so that in a fluid file we build a `Paragraph` like this:
```
Paragraph [ 
Text "Within each scenario bar plot, the bars represent: total warming ", Text (explainBars total "Total",)
Text ", warming contributions from CO2 ", Text (explainBars co2 "CO2"), Text " and from non-CO2 GHG's ", Text (explainBars nonco2 "Non-CO2")]
```

**Comment by rolyp** [comment]
Yes. Technically an untyped language would allow you to freely mix strings with `Link` constructors in the same collection, and so from that point of view we could omit `Text` entirely, but to give a type to such a collection would require union types. Staying within the spirit of “data types” feels better for now, and just means #1187 bumps up the priority list.

**Comment by JosephBond** [comment]
Okay, I've almost completed that segment of the task then. The constructor `Text` is already used in the graphics library for captions

**Comment by rolyp** [comment]
@JosephBond Thought a bit more about this. For the query underlying a `Link` to a substructure of a computed visualisation $v$ to properlyz select the desired subcomponent (e.g. a bar in a bar chart), I think we need for the `Link` to be “downstream of” $v$ in the pipeline. Otherwise we will end up with a secondary selection rather than a primary.

This won’t be what we end up with when you complete this initial task, because the text and the bar chart are “siblings” in a `MultiView`.  So this observation may be a useful constraint that will help us understand what to focus on text, as we’ll want to arrange for the bar chart to be the output $v$ and for the accompanying `Paragraph` values to somehow come later (as some kind of metadata able to refer to $v$).

(This issue is probably complicated by the sharing of value vertices in the _term graph_ representing values, but that’s perhaps more an artifact of the current design, and may change as we revisit the graph semantics).

**Comment by rolyp** [comment]
Given the observation from this morning, maybe the first pass over this should just take a string expression (rather function which returns a string) as the second argument.

**Comment by JosephBond** [comment]
Basic version of the `Link` constructor works, although it's sensitive to the shape of the value in the first argument. For example:
```
let total = map getTotal [ssp119, ssp245];
    meanTotal = (sum (map stackedBarHeight total)) / (length total)
in Paragraph [ ...; Link meanTotal " total warming (°C; total bar)"; ... ]
```
Works nicely, and lets you perform a selection like this:

![Image](https://github.com/user-attachments/assets/fc22b2c4-0460-47a9-8075-4c0e3825dd79)
But if you change the link to:
```
let total = map getTotal [ssp119, ssp245];
    meanTotal = (sum (map stackedBarHeight total)) / (length total)
in Paragraph [ ...; Link total " total warming (°C; total bar)"; ... ]
```
Then selections have no impact (because of the naive way I implemented the thing that gets/sets the selection on Links).

I know we discussed that we want to completely select everything upstream of the Link's first argument, but as yet I'm uncertain of the best way to move forward on that part of the task

**Comment by JosephBond** [comment]
Regarding the overlapping instances error:
It was hiding an issue where we did indeed need apply for `SelState`'s. Removing that original instance defining `MeetSemilattice` for datatypes that have `Functor` and `Apply` means we need to go around defining that instance for a bunch of types, like `Env` and `Val`. Therefore, I propose that we move SelStates to their own module which is imported by Lattice, which should in theory allow us to add the `MeetSemilattice` instance for `SelState` to the instance chain, whilst also auto-deriving `MeetSemilattice` for all other types that have `Functor` and `Apply`.

Edit:
Actually, this isn't quite as simple as I thought, we rely on the lattice constraints for a lot of utility functions on `SelState 𝕊`'s, so we would need a module that defines the `SelState` types, and non-lattice typeclasses, then in `App.Util` the functions that operate on them.

**Comment by rolyp** [comment]
Ok, let’s not worry about trying to complete this in time for the meeting – we can pick it up afterwards.
